flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectOfTypeStrategy.java

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * Type strategy for the {@code OBJECT_OF} function that infers the output type as a structured
+ * type.
+ *
+ * <p>This strategy creates a {@link DataTypes#STRUCTURED} type based on the provided class name and
+ * key-value pairs. The function signature is: {@code OBJECT_OF(className, key1, value1, key2,
+ * value2, ...)}
+ *
+ * <p>The strategy performs the following operations:
+ *
+ * <ul>
+ *   <li>Extracts the class name from the first argument (must be a string literal)
+ *   <li>Processes key-value pairs starting from the second argument
+ *   <li>Extracts field names from odd-positioned arguments (indices 1, 3, 5, ...)
+ *   <li>Normalizes field types from even-positioned arguments (indices 2, 4, 6, ...)
+ *   <li>Creates a structured type with the given class name and normalized fields
+ * </ul>
+ *
+ * <p>Field type normalization includes:
+ *
+ * <ul>
+ *   <li>Converting CHARACTER_STRING types (CHAR, VARCHAR) to nullable VARCHAR (STRING)
+ *   <li>Making all field types nullable for flexibility in structured types
+ * </ul>
+ *
+ * <p>The strategy returns {@code Optional.empty()} if the class name argument is not available as a
+ * literal value (e.g., during type inference testing scenarios).
+ *
+ * <p><b>Examples:</b>
+ *
+ * <ul>
+ *   <li>{@code OBJECT_OF('com.example.User', 'name', 'Alice', 'age', 30)} → {@code
+ *       STRUCTURED<com.example.User>(name STRING, age INT)}
+ *   <li>{@code OBJECT_OF('com.example.Point', 'x', 1.5, 'y', 2.0)} → {@code
+ *       STRUCTURED<com.example.Point>(x DOUBLE, y DOUBLE)}
+ * </ul>
+ *
+ * <p><b>Implementation Notes:</b>
+ *
+ * <ul>
+ *   <li>Field names must be available as string literals during type inference
+ *   <li>The class name is used for type identification but the runtime representation is RowData
+ *   <li>Uses {@link IntStream} for efficient processing of key-value pairs
+ * </ul>
+ *
+ * @see org.apache.flink.table.functions.BuiltInFunctionDefinitions#OBJECT_OF
+ * @see ObjectOfInputTypeStrategy
+ */
+public class ObjectOfTypeStrategy implements TypeStrategy {
+
+    private static DataType toStructuredType(
+            final String className, final CallContext callContext) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+        final DataTypes.Field[] fields =
+                IntStream.iterate(1, i -> i < argumentDataTypes.size(), i -> i + 2)
+                        .mapToObj(keyIdx -> toFieldDataType(callContext, keyIdx))
+                        .toArray(DataTypes.Field[]::new);
+
+        return DataTypes.STRUCTURED(className, fields);
+    }
+
+    private static Field toFieldDataType(final CallContext callContext, final int keyIdx) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+
+        final String fieldName =
+                callContext
+                        .getArgumentValue(keyIdx, String.class)
+                        .orElseThrow(IllegalStateException::new);
```

Comment:
probably better to call `CallContext#fail`

Reply:
> Therefore, this case should (ideally) never occur, and if so, it should be an IllegalState.

i disagree
the simplest way to make it happen
```sql
SELECT OBJECT_OF('class_name', CAST(NULL AS STRING), 'as');
```
And then user will see just `IllegalStateException` without any idea what to do about it

Reply:
i tend to think that classname and all keys should be always non nullable